plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php

```diff
@@ -275,6 +286,11 @@ protected function prepare_object_for_database( $request, $creating = false ) {
 			return new WP_Error( 'woocommerce_rest_empty_coupon_code', sprintf( __( 'The coupon code cannot be empty.', 'woocommerce' ), 'code' ), array( 'status' => 400 ) );
 		}
 
+		// Handle setting discount type first because that is needed for validation.
+		if ( isset( $request['discount_type'] ) ) {
+			$coupon->set_discount_type( $request['discount_type'] );
+		}
```

Comment:
This is the important relevant change for the fix (besides tests). The rest is from autoformatting.